File: 프로그래머스/lv2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    vector<vector<int>> matrix = vector(rows, vector(columns, 0));
    
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            matrix[i][j] = i * columns + j + 1;
        }
    }
    
    for (int i = 0; i < queries.size(); i++) {
        int small_x = queries[i][0] - 1;
        int small_y = queries[i][1] - 1;
        int large_x = queries[i][2] - 1;
        int large_y = queries[i][3] - 1;
        
        int x = small_x;
        int y = small_y;
        int init_value = matrix[small_x][small_y];
        int min_value = rows * columns + 1;

        while (true) {
            if (y == small_y + 1 && x == small_x) {
                matrix[x][y] = init_value;
                min_value = min(min_value, matrix[x][y]);
                break;
            } else if (y == small_y && x < large_x) {
                matrix[x][y] = matrix[x + 1][y];
                min_value = min(min_value, matrix[x][y]);
                x++;
            } else if (y < large_y && x == large_x) {
                matrix[x][y] = matrix[x][y + 1];
                min_value = min(min_value, matrix[x][y]);
                y++;
            } else if (y == large_y && x > small_x) {
                matrix[x][y] = matrix[x - 1][y];
                min_value = min(min_value, matrix[x][y]);
                x--;
            } else {
                matrix[x][y] = matrix[x][y - 1];
                min_value = min(min_value, matrix[x][y]);
                y--;
            }
        }
        
        answer.push_back(min_value);
    }

    
    return answer;
}

File: 프로그래머스/lv2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    square = [[j * columns + i + 1 for i in range(columns)] for j in range(rows)]
    answer = []
    
    for query in queries:
        x1, y1, x2, y2 = [q - 1 for q in query]
        
        mini = 10000
        mini = min(mini, square[x1][y1])
        store = square[x1][y1]
        
        x, y = x1 + 1, y1
        while not (x == x1 and y == y1):
            mini = min(mini, square[x][y])
            if y == y1:
                square[x - 1][y] = square[x][y]
                if x + 1 <= x2:
                    x += 1
                else:
                    y += 1
            elif x == x2:
                square[x][y - 1] = square[x][y]
                if y + 1 <= y2:
                    y += 1
                else:
                    x -= 1
            elif y == y2:
                square[x + 1][y] = square[x][y]
                if x - 1 >= x1:
                    x -= 1
                else:
                    y -= 1
            elif x == x1:
                square[x][y + 1] = square[x][y]
                if y - 1 >= y1:
                    y -= 1
        square[x1][y1 + 1] = store        
        answer.append(mini)
    
    
    return answer